### PR TITLE
fix(sync_file): return error to handlerFailureNoLock

### PR DESCRIPTION
**Pull request by FrankYang0529**
ref: https://github.com/longhorn/longhorn/issues/4852, https://github.com/harvester/harvester/issues/3123

**Test Step**

1. Install LH and set the `default-backing-image-manager-image` setting to this branch.
2. Create a BackingImage with a invalid checksum. We can see the `state` of the BackingImage will get into `failed`.

### CODE REVIEW

**Review by shuo-wu — COMMENTED** [review]
Can the below modification for [this](https://github.com/longhorn/backing-image-manager/blob/0435b5f5853e8c33c8830ec26690f1a2d9a167a0/pkg/sync/sync_file.go#L472-L474) fix this issue?
```
	defer func() {
		err = sf.finishProcessing(err)
	}()
```

**Comment by FrankYang0529** [comment]
> Can the below modification for [this](https://github.com/longhorn/backing-image-manager/blob/0435b5f5853e8c33c8830ec26690f1a2d9a167a0/pkg/sync/sync_file.go#L472-L474) fix this issue?
> 
> ```
> 	defer func() {
> 		err = sf.finishProcessing(err)
> 	}()
> ```

I used another branch to test this modification, but it doesn't work. The root cause is that we update `state` in [`handleFailureNoLock` function](https://github.com/longhorn/backing-image-manager/blob/e5911f81512b591e2fec671d30761c86c85f430a/pkg/sync/sync_file.go#L681-L701), so we need to give a correct error to it.

https://github.com/FrankYang0529/backing-image-manager/commit/a4a5f9081e6a56c390848fc056eb689c40c367a0

**Comment by shuo-wu** [comment]
OK, I see.
In my previous suggestion code, [the defer function `defer sf.handleFailureNoLock`](https://github.com/longhorn/backing-image-manager/blob/e5911f81512b591e2fec671d30761c86c85f430a/pkg/sync/sync_file.go#L609) inside `finishProcessingNoLock` does not get a correct error input. If the returned error of `finishProcessingNoLock` is named `finalErr`, then the defer function should be `defer sf.handleFailureNoLock(finalErr)`.

If you prefer the current modification, it looks good as well but with one minor suggestion: `finishProcessingNoLock` does not need to receive an error input. Instead, the error can be handled in the caller part.

**Comment by FrankYang0529** [comment]
Hi @shuo-wu, thanks for the suggestion. After reading https://trstringer.com/golang-deferred-function-error-handling/, I updated `Fetch`, `DownloadFromURL`, `IdleTimeoutCopyToFile`, and `Receive` to handle errors in the `sf.finishProcessing` function.

File: pkg/sync/sync_file.go [hunk 1]
```diff
@@ -424,7 +425,9 @@ func (sf *SyncingFile) Fetch(srcFilePath string) (err error) {
 	}
 
 	defer func() {
-		sf.finishProcessing(err)
+		if finalErr := sf.finishProcessing(err); finalErr != nil {
+			err = finalErr
+		}
```
Comment (shuo-wu) [inline pkg/sync/sync_file.go:430]: NIT: `err = sf.finishProcessing(err)`

**Review by shuo-wu — APPROVED** [review]
LGTM

**Review by innobead — APPROVED** [review]

**Pull request merged**